src/txmempool.cpp

```diff
@@ -57,46 +57,75 @@ size_t CTxMemPoolEntry::GetTxSize() const
 // Update the given tx for any in-mempool descendants.
 // Assumes that setMemPoolChildren is correct for the given tx and all
 // descendants.
-void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap &cachedDescendants, const std::set<uint256> &setExclude)
+void CTxMemPool::UpdateForDescendants(txiter update_it, cacheMap& cache)
 {
-    setEntries stageEntries, setAllDescendants;
-    stageEntries = GetMemPoolChildren(updateIt);
-
-    while (!stageEntries.empty()) {
-        const txiter cit = *stageEntries.begin();
-        setAllDescendants.insert(cit);
-        stageEntries.erase(cit);
-        const setEntries &setChildren = GetMemPoolChildren(cit);
-        for (txiter childEntry : setChildren) {
-            cacheMap::iterator cacheIt = cachedDescendants.find(childEntry);
-            if (cacheIt != cachedDescendants.end()) {
-                // We've already calculated this one, just add the entries for this set
-                // but don't traverse again.
-                for (txiter cacheEntry : cacheIt->second) {
-                    setAllDescendants.insert(cacheEntry);
+    const auto epoch = GetFreshEpoch();
+    // set up the new_cache_line to contain all of our transaction's children (note --
+    // already de-duplicated in case multiple outputs of ours are spent in one
+    // transaction).
+    // Always creates a new cache line entry.
+    vecEntries& new_cache_line = cache[update_it];
+    assert(new_cache_line.empty());
+    // next_idx_to_walk index keeps track of the elements that we've
+    // already expanded.
+    // Invariants:
+    // If index is < next_idx_to_walk, we've already walked it.
+    // If index is >= next_idx_to_walk, we need to walk it.
+    // If next_idx_to_walk >= new_cache_line.size(), we're finished.
+    size_t next_idx_to_walk = 0;
+    txiter next_it = update_it;
+    do {
+        for (const txiter child_it : GetMemPoolChildren(next_it)) {
```

Comment:
```suggestion
        for (const txiter& child_it : GetMemPoolChildren(next_it)) {
```

to avoid these compile warnings:
```
txmempool.cpp:78:27: warning: loop variable 'child_it' of type 'const CTxMemPool::txiter' (aka 'const hashed_index_iterator<hashed_index_node<ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, index_node_base<CTxMemPoolEntry, std::allocator<CTxMemPoolEntry> > > > >, boost::multi_index::detail::hashed_unique_tag>, bucket_array<std::allocator<CTxMemPoolEntry> >, boost::multi_index::detail::hashed_index_global_iterator_tag>') creates a copy from type 'const CTxMemPool::txiter' [-Wrange-loop-analysis]
        for (const txiter child_it : GetMemPoolChildren(next_it)) {
                          ^
txmempool.cpp:78:14: note: use reference type 'const CTxMemPool::txiter &' (aka 'const hashed_index_iterator<hashed_index_node<ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, ordered_index_node<boost::multi_index::detail::null_augment_policy, index_node_base<CTxMemPoolEntry, std::allocator<CTxMemPoolEntry> > > > >, boost::multi_index::detail::hashed_unique_tag>, bucket_array<std::allocator<CTxMemPoolEntry> >, boost::multi_index::detail::hashed_index_global_iterator_tag> &') to prevent copying
        for (const txiter child_it : GetMemPoolChildren(next_it)) {
```

Reply:
Suppose you compiled with clang?

Reply:
Yes, this was a clang warning (I often compile with both gcc and clang for anything non-trivial as they seem to have complementary results).

Reply:
Hmm I think specifically we *do* want copying here because it's an iter, which in boost::multiindex should be just a pointer, so a reference to an iter implies extra unnecessary indirection?

@sipa is probably most familiar with multiindex innards to know if this is correct.

Reply:
Yeah, `CTxMemPool::txiter` is just a single pointer, a copy would be desirable here.

However, I suspect that the compiler will figure this out, and produce pretty much equal code for both.

Reply:
Thanks sipa.

My preference is to leave the code as is if the behavior is correct & desirable (even if the compiler is likely to optimize out the indirection), as the warning seems to be just an aggressive warning from clang, unless a warning is excessively bad. The right solution seems to be to annotate somehow that txiter is a copyable type. How can this be done?

Reply:
It seems there are two solutions: not mark the loop variable const (if you want something mutable, a copy is needed anyway), or upgrade to clang 10.

Reply:
FWIW, at -O2, clang 9 compiles both to the same code.

Reply:
txiter is already const (it would be illegal to modify a mempool entry by the iterator). I don't think we write 'const txiter' anywhere

Reply:
It's a const_iterator. The `const` here just prevents modifying the txiter object, not what it points to. It's a vaguely sensible thing to do when you know a variable isn't going to be modified.

Reply:
@MarcoFalke this is not quite true, it's perfectly legal to modify through the iterator via at least two means:

1) Const cast -- the txiter is known to be non const *somewhere* so updating is fine
2) Mutable fields -- the txiter epochs are mutable, same with the vtxhashesidx

In this case we are modifying the epochs via the mutable fields even though it's const.

Reply:
On Clang 6 (Debian), I confirm that passing by value and making it non-const also removes the warning.

If the resulting code is the same, my suggestion would be to appease the warnings to not add needlessly to the noise and friction for reviewers.

Reply:
Yes, on Clang 6 (Debian), making the txiters non-const also removes the warning.

If the resulting code is the same, my suggestion would be to appease the warnings to not add needlessly to the noise and friction for reviewers, here and down the road. Could be added as a follow-up if this is merged.